File: tools/portable-validator/validator_governance.go
package main

import (
	"crypto/ed25519"
	"crypto/x509"
	"encoding/base64"
	"errors"
	"fmt"
	"sort"
)

const (
	validatorGovernanceProfile = "STRATUM-VALIDATOR-GOVERNANCE/1"
	validatorSetChangeVersion  = "STRATUM-VALIDATOR-SET-CHANGE/1"
	validatorSetChangeDomain   = "STRATUM/VALIDATOR_SET_CHANGE/1"
	validatorSetStateProfile   = "STRATUM-VALIDATOR-SET-STATE/1"
	validatorSetStateDomain    = "STRATUM/VALIDATOR_SET_STATE/1"
	validatorSubjectDomain     = "STRATUM/VALIDATOR/SUBJECT/1"
)

type ValidatorGovernanceMember struct {
	AuthorityMemberID string `json:"authorityMemberId"`
	KeyID             string `json:"keyId"`
	Purpose           string `json:"purpose"`
	Algorithm         string `json:"algorithm"`
	PublicKeyDerB64   string `json:"publicKeyDerB64"`
	State             string `json:"state"`
	ValidFromHeight   int64  `json:"validFromHeight"`
	ValidUntilHeight  *int64 `json:"validUntilHeight"`
}

type ValidatorGovernancePolicy struct {
	Profile              string                      `json:"profile"`
	ChainID              string                      `json:"chainId"`
	AuthorityID          string                      `json:"authorityId"`
	AuthorityType        string                      `json:"authorityType"`
	OrganizationID       string                      `json:"organizationId"`
	Jurisdiction         *string                     `json:"jurisdiction"`
	Scope                []string                    `json:"scope"`
	VotingPolicy         string                      `json:"votingPolicy"`
	Threshold            int                         `json:"threshold"`
	EffectiveFromHeight  int64                       `json:"effectiveFromHeight"`
	EffectiveUntilHeight *int64                      `json:"effectiveUntilHeight"`
	Status               string                      `json:"status"`
	Members              []ValidatorGovernanceMember `json:"members"`
}

type ValidatorSetChangeAction struct {
	Domain                   string `json:"domain"`
	ActionVersion            string `json:"actionVersion"`
	ChainID                  string `json:"chainId"`
	ActionID                 string `json:"actionId"`
	ActionType               string `json:"actionType"`
	ValidatorID              string `json:"validatorId"`
	ApprovedAtHeight         int64  `json:"approvedAtHeight"`
	EffectiveHeight          int64  `json:"effectiveHeight"`
	PreviousValidatorSetRoot string `json:"previousValidatorSetRoot"`
	NextValidatorSetRoot     string `json:"nextValidatorSetRoot"`
	PreviousValidatorSetHash string `json:"previousValidatorSetHash"`
	NextValidatorSetHash     string `json:"nextValidatorSetHash"`
	SubjectIdentityHash      string `json:"subjectIdentityHash"`
	GovernancePolicyHash     string `json:"governancePolicyHash"`
	ReasonHash               string `json:"reasonHash"`
}

type ValidatorSetChangeSignature struct {
	AuthorityMemberID string `json:"authorityMemberId"`
	KeyID             string `json:"keyId"`
	Algorithm         string `json:"algorithm"`
	Domain            string `json:"domain"`
	SignatureB64      string `json:"signatureB64"`
}

type ValidatorSetChangeProof struct {
	Action               ValidatorSetChangeAction      `json:"action"`
	GovernancePolicy     ValidatorGovernancePolicy     `json:"governancePolicy"`
	PreviousValidatorSet SnapshotValidatorSet          `json:"previousValidatorSet"`
	NextValidatorSet     SnapshotValidatorSet          `json:"nextValidatorSet"`
	Signatures           []ValidatorSetChangeSignature `json:"signatures"`
}

type ValidatorSetChangeVerification struct {
	Valid                    bool     `json:"valid"`
	ActionID                 string   `json:"actionId"`
	ActionType               string   `json:"actionType"`
	ValidatorID              string   `json:"validatorId"`
	ApprovedAtHeight         int64    `json:"approvedAtHeight"`
	EffectiveHeight          int64    `json:"effectiveHeight"`
	GovernancePolicyHash     string   `json:"governancePolicyHash"`
	PreviousValidatorSetRoot string   `json:"previousValidatorSetRoot"`
	NextValidatorSetRoot     string   `json:"nextValidatorSetRoot"`
	PreviousValidatorSetHash string   `json:"previousValidatorSetHash"`
	NextValidatorSetHash     string   `json:"nextValidatorSetHash"`
	EligibleAuthorityCount   int      `json:"eligibleAuthorityCount"`
	RequiredAuthority        int      `json:"requiredAuthority"`
	ValidSigners             []string `json:"validSigners"`
}

func governanceMemberEligibleAtHeight(member ValidatorGovernanceMember, height int64) bool {
	return member.State == "ACTIVE" && member.ValidFromHeight <= height && (member.ValidUntilHeight == nil || *member.ValidUntilHeight > height)
}

func requiredGovernanceSupermajority(n int) (int, error) {
	if n < 1 {
		return 0, errors.New("validator governance requires at least one eligible authority member")
	}
	return (2*n)/3 + 1, nil
}

func canonicalGovernancePolicyHash(policy ValidatorGovernancePolicy) (string, error) {
	copyPolicy := policy
	copyPolicy.Scope = append([]string(nil), policy.Scope...)
	sort.Strings(copyPolicy.Scope)
	copyPolicy.Members = append([]ValidatorGovernanceMember(nil), policy.Members...)
	sort.Slice(copyPolicy.Members, func(i, j int) bool {
		if copyPolicy.Members[i].AuthorityMemberID == copyPolicy.Members[j].AuthorityMemberID {
			return copyPolicy.Members[i].KeyID < copyPolicy.Members[j].KeyID
		}
		return copyPolicy.Members[i].AuthorityMemberID < copyPolicy.Members[j].AuthorityMemberID
	})
	return canonicalHashValue(copyPolicy)
}

func validatorSetStateHash(set SnapshotValidatorSet) (string, error) {
	members := append([]SnapshotValidator(nil), set.Members...)
	for i := range members {
		members[i].Keys = append([]SnapshotConsensusKey(nil), members[i].Keys...)
		sort.Slice(members[i].Keys, func(a, b int) bool { return members[i].Keys[a].KeyID < members[i].Keys[b].KeyID })
	}
	sort.Slice(members, func(i, j int) bool { return members[i].ValidatorID < members[j].ValidatorID })
	return canonicalHashValue(map[string]any{"domain": validatorSetStateDomain, "profile": validatorSetStateProfile, "setVersion": set.SetVersion, "chainId": set.ChainID, "members": members})
}

func validatorSubjectIdentityHash(member SnapshotValidator) (string, error) {
	return canonicalHashValue(map[string]any{"domain": validatorSubjectDomain, "validatorId": member.ValidatorID, "identityUuid": member.IdentityUUID, "operatorOrg": member.OperatorOrg})
}

func governanceRequiredScope(actionType string) string {
	if actionType == "ROTATE_CONSENSUS_KEY" {
		return "VALIDATOR_CONSENSUS_KEYS"
	}
	return "VALIDATOR_MEMBERSHIP"
}

func validatorMemberByID(set SnapshotValidatorSet, id string) (SnapshotValidator, bool) {
	for _, member := range set.Members {
		if member.ValidatorID == id {
			return member, true
		}
	}
	return SnapshotValidator{}, false
}

func sameValidatorIdentity(a, b SnapshotValidator) bool {
	return a.ValidatorID == b.ValidatorID && a.IdentityUUID == b.IdentityUUID && a.OperatorOrg == b.OperatorOrg
}

func assertOnlyGovernedTargetChanged(previous, next SnapshotValidatorSet, targetID string) error {
	all := map[string]bool{}
	for _, member := range previous.Members { all[member.ValidatorID] = true }
	for _, member := range next.Members { all[member.ValidatorID] = true }
	for id := range all {
		if id == targetID { continue }
		a, okA := validatorMemberByID(previous, id); b, okB := validatorMemberByID(next, id)
		if !okA || !okB { return fmt.Errorf("unrelated validator %s changed in validator-set action", id) }
		ha, err := canonicalHashValue(a); if err != nil { return err }
		hb, err := canonicalHashValue(b); if err != nil { return err }
		if ha != hb { return fmt.Errorf("unrelated validator %s changed in validator-set action", id) }
	}
	return nil
}

func assertGovernedMembershipTransition(proof ValidatorSetChangeProof) (SnapshotValidator, error) {
	action := proof.Action
	before, hasBefore := validatorMemberByID(proof.PreviousValidatorSet, action.ValidatorID)
	after, hasAfter := validatorMemberByID(proof.NextValidatorSet, action.ValidatorID)
	priorHeight := action.EffectiveHeight - 1
	if action.ActionType == "ACTIVATE" {
		if hasBefore { return SnapshotValidator{}, errors.New("ACTIVATE v1 requires target absent from previous canonical set") }
		if !hasAfter || after.ActivationHeight != action.EffectiveHeight || after.RetirementHeight != nil || !snapshotValidatorActive(after, action.EffectiveHeight) { return SnapshotValidator{}, errors.New("ACTIVATE does not produce ACTIVE target at effective height") }
		if _, err := activeSnapshotConsensusKey(after, action.EffectiveHeight); err != nil { return SnapshotValidator{}, err }
		return after, nil
	}
	if !hasBefore || !hasAfter { return SnapshotValidator{}, fmt.Errorf("%s requires target in both canonical set snapshots", action.ActionType) }
	if !sameValidatorIdentity(before, after) { return SnapshotValidator{}, errors.New("validator permanent identity changed across governed action") }
	switch action.ActionType {
	case "QUARANTINE", "RETIRE", "REVOKE":
		if !snapshotValidatorActive(before, priorHeight) { return SnapshotValidator{}, fmt.Errorf("%s target was not ACTIVE before effective height", action.ActionType) }
		if snapshotValidatorActive(after, action.EffectiveHeight) || after.RetirementHeight == nil || *after.RetirementHeight != action.EffectiveHeight { return SnapshotValidator{}, fmt.Errorf("%s does not remove target from ACTIVE set at effective height", action.ActionType) }
		return after, nil
	case "REINSTATE":
		if snapshotValidatorActive(before, priorHeight) { return SnapshotValidator{}, errors.New("REINSTATE target was already ACTIVE before effective height") }
		if after.ActivationHeight != action.EffectiveHeight || after.RetirementHeight != nil || !snapshotValidatorActive(after, action.EffectiveHeight) { return SnapshotValidator{}, errors.New("REINSTATE does not restore target at effective height") }
		if _, err := activeSnapshotConsensusKey(after, action.EffectiveHeight); err != nil { return SnapshotValidator{}, err }
		return after, nil
	default:
		return SnapshotValidator{}, fmt.Errorf("unsupported membership transition %s", action.ActionType)
	}
}

func assertGovernedConsensusKeyRotation(proof ValidatorSetChangeProof) (SnapshotValidator, error) {
	action := proof.Action
	before, okBefore := validatorMemberByID(proof.PreviousValidatorSet, action.ValidatorID)
	after, okAfter := validatorMemberByID(proof.NextValidatorSet, action.ValidatorID)
	if !okBefore || !okAfter || !sameValidatorIdentity(before, after) { return SnapshotValidator{}, errors.New("ROTATE_CONSENSUS_KEY requires same permanent validator identity") }
	if before.ActivationHeight != after.ActivationHeight { return SnapshotValidator{}, errors.New("consensus-key rotation must not change validator activation height") }
	if (before.RetirementHeight == nil) != (after.RetirementHeight == nil) { return SnapshotValidator{}, errors.New("consensus-key rotation must not change validator retirement state") }
	if before.RetirementHeight != nil && *before.RetirementHeight != *after.RetirementHeight { return SnapshotValidator{}, errors.New("consensus-key rotation must not change validator retirement height") }
	if !snapshotValidatorActive(before, action.EffectiveHeight-1) || !snapshotValidatorActive(after, action.EffectiveHeight) { return SnapshotValidator{}, errors.New("consensus-key rotation requires validator ACTIVE across effective height") }
	oldActive, err := activeSnapshotConsensusKey(before, action.EffectiveHeight-1); if err != nil { return SnapshotValidator{}, err }
	newActive, err := activeSnapshotConsensusKey(after, action.EffectiveHeight); if err != nil { return SnapshotValidator{}, err }
	if oldActive.KeyID == newActive.KeyID { return SnapshotValidator{}, errors.New("consensus-key rotation did not change active key") }
	beforeKeys := map[string]SnapshotConsensusKey{}; afterKeys := map[string]SnapshotConsensusKey{}
	for _, key := range before.Keys { beforeKeys[key.KeyID] = key }
	for _, key := range after.Keys { afterKeys[key.KeyID] = key }
	retiredOld, ok := afterKeys[oldActive.KeyID]
	if !ok || retiredOld.PublicKeyDerB64 != oldActive.PublicKeyDerB64 || retiredOld.RetiredAtHeight == nil || *retiredOld.RetiredAtHeight != action.EffectiveHeight { return SnapshotValidator{}, errors.New("previous CONSENSUS key must retire exactly at the effective height") }
	newIDs := []string{}
	for keyID := range afterKeys { if _, exists := beforeKeys[keyID]; !exists { newIDs = append(newIDs, keyID) } }
	if len(newIDs) != 1 || newIDs[0] != newActive.KeyID || newActive.ActiveFromHeight != action.EffectiveHeight { return SnapshotValidator{}, errors.New("exactly one replacement CONSENSUS key must activate at effective height") }
	for keyID, key := range beforeKeys {
		if keyID == oldActive.KeyID { continue }
		nextKey, ok := afterKeys[keyID]; if !ok { return SnapshotValidator{}, fmt.Errorf("unrelated key history %s changed during rotation", keyID) }
		ha, err := canonicalHashValue(key); if err != nil { return SnapshotValidator{}, err }
		hb, err := canonicalHashValue(nextKey); if err != nil { return SnapshotValidator{}, err }
		if ha != hb { return SnapshotValidator{}, fmt.Errorf("unrelated key history %s changed during rotation", keyID) }
	}
	return after, nil
}

func verifyValidatorSetChangeProof(proof ValidatorSetChangeProof, expectedChainID, expectedGovernancePolicyHash, expectedPreviousValidatorSetRoot string) (ValidatorSetChangeVerification, error) {
	action := proof.Action; policy := proof.GovernancePolicy; previous := proof.PreviousValidatorSet; next := proof.NextValidatorSet
	if action.Domain != validatorSetChangeDomain || action.ActionVersion != validatorSetChangeVersion { return ValidatorSetChangeVerification{}, errors.New("unsupported validator-set change domain/version") }
	if policy.Profile != validatorGovernanceProfile || policy.AuthorityType != "VALIDATOR" || policy.VotingPolicy != "THRESHOLD" { return ValidatorSetChangeVerification{}, errors.New("unsupported validator governance policy profile") }
	if action.ChainID != expectedChainID || policy.ChainID != expectedChainID || previous.ChainID != expectedChainID || next.ChainID != expectedChainID { return ValidatorSetChangeVerification{}, errors.New("validator governance chainId mismatch") }
	if action.EffectiveHeight <= action.ApprovedAtHeight { return ValidatorSetChangeVerification{}, errors.New("validator-set changes must activate after approval height") }
	if policy.Status != "ACTIVE" || policy.EffectiveFromHeight > action.ApprovedAtHeight || (policy.EffectiveUntilHeight != nil && *policy.EffectiveUntilHeight <= action.ApprovedAtHeight) { return ValidatorSetChangeVerification{}, errors.New("validator governance policy was not ACTIVE at approval height") }
	requiredScope := governanceRequiredScope(action.ActionType); hasScope := false
	for _, scope := range policy.Scope { if scope == requiredScope { hasScope = true } }
	if !hasScope { return ValidatorSetChangeVerification{}, fmt.Errorf("validator governance policy lacks %s authority", requiredScope) }
	policyHash, err := canonicalGovernancePolicyHash(policy); if err != nil { return ValidatorSetChangeVerification{}, err }
	if policyHash != expectedGovernancePolicyHash || action.GovernancePolicyHash != policyHash { return ValidatorSetChangeVerification{}, errors.New("validator governance policy hash does not match independently trusted policy") }
	previousHash, err := validatorSetStateHash(previous); if err != nil { return ValidatorSetChangeVerification{}, err }
	nextHash, err := validatorSetStateHash(next); if err != nil { return ValidatorSetChangeVerification{}, err }
	if action.PreviousValidatorSetHash != previousHash || action.NextValidatorSetHash != nextHash { return ValidatorSetChangeVerification{}, errors.New("validator-set state hash mismatch") }
	previousRoot, err := snapshotValidatorSetRoot(previous, action.EffectiveHeight-1); if err != nil { return ValidatorSetChangeVerification{}, err }
	if previousRoot != expectedPreviousValidatorSetRoot || action.PreviousValidatorSetRoot != previousRoot { return ValidatorSetChangeVerification{}, errors.New("previous validator-set root does not match independently trusted canonical root") }
	nextRoot, err := snapshotValidatorSetRoot(next, action.EffectiveHeight); if err != nil { return ValidatorSetChangeVerification{}, err }
	if action.NextValidatorSetRoot != nextRoot { return ValidatorSetChangeVerification{}, errors.New("next validator-set root mismatch") }
	if err := assertOnlyGovernedTargetChanged(previous, next, action.ValidatorID); err != nil { return ValidatorSetChangeVerification{}, err }
	var subject SnapshotValidator
	if action.ActionType == "ROTATE_CONSENSUS_KEY" { subject, err = assertGovernedConsensusKeyRotation(proof) } else { subject, err = assertGovernedMembershipTransition(proof) }
	if err != nil { return ValidatorSetChangeVerification{}, err }
	subjectHash, err := validatorSubjectIdentityHash(subject); if err != nil { return ValidatorSetChangeVerification{}, err }
	if action.SubjectIdentityHash != subjectHash { return ValidatorSetChangeVerification{}, errors.New("validator subject identity hash mismatch") }
	for label, value := range map[string]string{"governancePolicyHash":action.GovernancePolicyHash,"previousValidatorSetRoot":action.PreviousValidatorSetRoot,"nextValidatorSetRoot":action.NextValidatorSetRoot,"previousValidatorSetHash":action.PreviousValidatorSetHash,"nextValidatorSetHash":action.NextValidatorSetHash,"subjectIdentityHash":action.SubjectIdentityHash,"reasonHash":action.ReasonHash} { if !isSHA256(value) { return ValidatorSetChangeVerification{}, fmt.Errorf("%s must be lowercase SHA-256", label) } }

	eligible := map[string]ValidatorGovernanceMember{}
	for _, member := range policy.Members {
		if governanceMemberEligibleAtHeight(member, action.ApprovedAtHeight) {
			if _, exists := eligible[member.AuthorityMemberID]; exists { return ValidatorSetChangeVerification{}, fmt.Errorf("duplicate governance authority member %s", member.AuthorityMemberID) }
			eligible[member.AuthorityMemberID] = member
		}
	}
	supermajority, err := requiredGovernanceSupermajority(len(eligible)); if err != nil { return ValidatorSetChangeVerification{}, err }
	required := supermajority; if policy.Threshold > required { required = policy.Threshold }
	if required > len(eligible) { return ValidatorSetChangeVerification{}, fmt.Errorf("validator governance threshold %d exceeds %d eligible authority members", required, len(eligible)) }
	normalizedAction, err := canonicalValue(action); if err != nil { return ValidatorSetChangeVerification{}, err }
	canonicalAction, err := canonicalJSON(normalizedAction); if err != nil { return ValidatorSetChangeVerification{}, err }
	payload := []byte(canonicalAction)
	valid := map[string]bool{}
	for _, sig := range proof.Signatures {
		if valid[sig.AuthorityMemberID] { continue }
		member, ok := eligible[sig.AuthorityMemberID]; if !ok || sig.KeyID != member.KeyID || sig.Algorithm != "Ed25519" || sig.Domain != validatorSetChangeDomain { continue }
		der, err := base64.StdEncoding.DecodeString(member.PublicKeyDerB64); if err != nil { continue }
		parsed, err := x509.ParsePKIXPublicKey(der); if err != nil { continue }
		pub, ok := parsed.(ed25519.PublicKey); if !ok { continue }
		signature, err := base64.StdEncoding.DecodeString(sig.SignatureB64); if err != nil { continue }
		if ed25519.Verify(pub, payload, signature) { valid[sig.AuthorityMemberID] = true }
	}
	validSigners := make([]string,0,len(valid)); for signer := range valid { validSigners=append(validSigners,signer) }; sort.Strings(validSigners)
	if len(validSigners) < required { return ValidatorSetChangeVerification{}, fmt.Errorf("validator governance supermajority not met: %d/%d; %d required", len(validSigners), len(eligible), required) }
	return ValidatorSetChangeVerification{true,action.ActionID,action.ActionType,action.ValidatorID,action.ApprovedAtHeight,action.EffectiveHeight,policyHash,previousRoot,nextRoot,previousHash,nextHash,len(eligible),required,validSigners},nil
}

File: lib/redbook/validator-governance.ts
import {createPublicKey,verify as verifySignature} from 'crypto';
import {canonicalHash,canonicalize} from '@/lib/server/hash';
import {activeConsensusKeyAtHeight,validatorActiveAtHeight,validatorSetRootAtHeight} from './snapshot-trust';
import {snapshotValidatorSetSchema,type SnapshotValidatorSet} from './schema/snapshot';
import {
 VALIDATOR_SET_CHANGE_DOMAIN,
 VALIDATOR_SET_STATE_DOMAIN,
 VALIDATOR_SET_STATE_PROFILE,
 VALIDATOR_SUBJECT_DOMAIN,
 validatorGovernancePolicySchema,
 validatorSetChangeProofSchema,
 type ValidatorGovernancePolicy,
 type ValidatorSetChangeProof,
} from './schema/validator-governance';

function governanceMemberEligibleAtHeight(member:ValidatorGovernancePolicy['members'][number],height:number){
 return member.state==='ACTIVE'&&member.validFromHeight<=height&&(member.validUntilHeight===null||member.validUntilHeight>height);
}

export function requiredGovernanceSupermajority(eligibleAuthorityCount:number){
 if(!Number.isInteger(eligibleAuthorityCount)||eligibleAuthorityCount<1)throw new Error('eligibleAuthorityCount must be a positive integer');
 return Math.floor((2*eligibleAuthorityCount)/3)+1;
}

export function canonicalGovernancePolicy(input:ValidatorGovernancePolicy|unknown){
 const policy=validatorGovernancePolicySchema.parse(input);
 return{
  ...policy,
  scope:[...policy.scope].sort(),
  members:[...policy.members].sort((a,b)=>a.authorityMemberId.localeCompare(b.authorityMemberId)||a.keyId.localeCompare(b.keyId)),
 };
}

export function governancePolicyHash(input:ValidatorGovernancePolicy|unknown){
 return canonicalHash(canonicalGovernancePolicy(input));
}

export function canonicalValidatorSetState(input:SnapshotValidatorSet|unknown){
 const set=snapshotValidatorSetSchema.parse(input);
 return{
  domain:VALIDATOR_SET_STATE_DOMAIN,
  profile:VALIDATOR_SET_STATE_PROFILE,
  setVersion:set.setVersion,
  chainId:set.chainId,
  members:[...set.members].map(member=>({...member,keys:[...member.keys].sort((a,b)=>a.keyId.localeCompare(b.keyId))})).sort((a,b)=>a.validatorId.localeCompare(b.validatorId)),
 };
}

export function validatorSetStateHash(input:SnapshotValidatorSet|unknown){
 return canonicalHash(canonicalValidatorSetState(input));
}

export function validatorSubjectIdentityHash(member:SnapshotValidatorSet['members'][number]){
 return canonicalHash({domain:VALIDATOR_SUBJECT_DOMAIN,validatorId:member.validatorId,identityUuid:member.identityUuid,operatorOrg:member.operatorOrg});
}

function requiredScope(actionType:ValidatorSetChangeProof['action']['actionType']){
 return actionType==='ROTATE_CONSENSUS_KEY'?'VALIDATOR_CONSENSUS_KEYS' as const:'VALIDATOR_MEMBERSHIP' as const;
}

function sameIdentity(a:SnapshotValidatorSet['members'][number],b:SnapshotValidatorSet['members'][number]){
 return a.validatorId===b.validatorId&&a.identityUuid===b.identityUuid&&a.operatorOrg===b.operatorOrg;
}

function byValidatorId(set:SnapshotValidatorSet){return new Map(set.members.map(member=>[member.validatorId,member]));}

function assertOnlyTargetChanged(previous:SnapshotValidatorSet,next:SnapshotValidatorSet,targetId:string){
 const before=byValidatorId(previous);const after=byValidatorId(next);
 const allIds=new Set([...before.keys(),...after.keys()]);
 for(const id of allIds){
  if(id===targetId)continue;
  const a=before.get(id);const b=after.get(id);
  if(!a||!b||canonicalHash(a)!==canonicalHash(b))throw new Error(`Unrelated validator ${id} changed in validator-set action`);
 }
}

function assertMembershipTransition(proof:ValidatorSetChangeProof){
 const {action,previousValidatorSet:previous,nextValidatorSet:next}=proof;
 const before=byValidatorId(previous).get(action.validatorId);
 const after=byValidatorId(next).get(action.validatorId);
 const priorHeight=action.effectiveHeight-1;
 if(action.actionType==='ACTIVATE'){
  if(before)throw new Error('ACTIVATE v1 requires the target validator to be absent from the previous canonical set');
  if(!after||after.activationHeight!==action.effectiveHeight||after.retirementHeight!==null||!validatorActiveAtHeight(after,action.effectiveHeight))throw new Error('ACTIVATE does not produce an ACTIVE target at the effective height');
  activeConsensusKeyAtHeight(after,action.effectiveHeight);
  return after;
 }
 if(!before||!after)throw new Error(`${action.actionType} requires the target validator in both canonical set snapshots`);
 if(!sameIdentity(before,after))throw new Error('Validator permanent identity changed across governed action');
 if(action.actionType==='QUARANTINE'||action.actionType==='RETIRE'||action.actionType==='REVOKE'){
  if(!validatorActiveAtHeight(before,priorHeight))throw new Error(`${action.actionType} target was not ACTIVE before the effective height`);
  if(validatorActiveAtHeight(after,action.effectiveHeight)||after.retirementHeight!==action.effectiveHeight)throw new Error(`${action.actionType} does not remove the target from the ACTIVE set at the effective height`);
  return after;
 }
 if(action.actionType==='REINSTATE'){
  if(validatorActiveAtHeight(before,priorHeight))throw new Error('REINSTATE target was already ACTIVE before the effective height');
  if(after.activationHeight!==action.effectiveHeight||after.retirementHeight!==null||!validatorActiveAtHeight(after,action.effectiveHeight))throw new Error('REINSTATE does not restore the target at the effective height');
  activeConsensusKeyAtHeight(after,action.effectiveHeight);
  return after;
 }
 throw new Error(`Unsupported membership transition ${action.actionType}`);
}

function assertConsensusKeyRotation(proof:ValidatorSetChangeProof){
 const {action,previousValidatorSet:previous,nextValidatorSet:next}=proof;
 const before=byValidatorId(previous).get(action.validatorId);const after=byValidatorId(next).get(action.validatorId);
 if(!before||!after||!sameIdentity(before,after))throw new Error('ROTATE_CONSENSUS_KEY requires the same permanent validator identity in both sets');
 if(before.activationHeight!==after.activationHeight||before.retirementHeight!==after.retirementHeight)throw new Error('Consensus-key rotation must not change validator membership lifecycle');
 if(!validatorActiveAtHeight(before,action.effectiveHeight-1)||!validatorActiveAtHeight(after,action.effectiveHeight))throw new Error('Consensus-key rotation requires the validator ACTIVE across the effective height');
 const oldActive=activeConsensusKeyAtHeight(before,action.effectiveHeight-1);
 const newActive=activeConsensusKeyAtHeight(after,action.effectiveHeight);
 if(oldActive.keyId===newActive.keyId)throw new Error('Consensus-key rotation did not change the active key');
 const beforeKeys=new Map(before.keys.map(key=>[key.keyId,key]));const afterKeys=new Map(after.keys.map(key=>[key.keyId,key]));
 const retiredOld=afterKeys.get(oldActive.keyId);
 if(!retiredOld||retiredOld.publicKeyDerB64!==oldActive.publicKeyDerB64||retiredOld.retiredAtHeight!==action.effectiveHeight)throw new Error('Previous CONSENSUS key must retire exactly at the effective height');
 const newlyAdded=[...afterKeys.keys()].filter(keyId=>!beforeKeys.has(keyId));
 if(newlyAdded.length!==1||newlyAdded[0]!==newActive.keyId||newActive.activeFromHeight!==action.effectiveHeight)throw new Error('Exactly one replacement CONSENSUS key must activate at the effective height');
 for(const [keyId,key] of beforeKeys){
  if(keyId===oldActive.keyId)continue;
  const nextKey=afterKeys.get(keyId);
  if(!nextKey||canonicalHash(key)!==canonicalHash(nextKey))throw new Error(`Unrelated key history ${keyId} changed during rotation`);
 }
 return after;
}

export function verifyValidatorSetChangeProof(args:{
 proof:unknown;
 expectedChainId:string;
 expectedGovernancePolicyHash:string;
 expectedPreviousValidatorSetRoot:string;
}){
 const proof=validatorSetChangeProofSchema.parse(args.proof);
 const {action,governancePolicy:policy,previousValidatorSet:previous,nextValidatorSet:next}=proof;
 if(action.chainId!==args.expectedChainId||policy.chainId!==args.expectedChainId||previous.chainId!==args.expectedChainId||next.chainId!==args.expectedChainId)throw new Error('Validator governance chainId mismatch');
 if(action.effectiveHeight<=action.approvedAtHeight)throw new Error('Validator-set changes must activate after their approval height');
 if(policy.status!=='ACTIVE'||policy.effectiveFromHeight>action.approvedAtHeight||(policy.effectiveUntilHeight!==null&&policy.effectiveUntilHeight<=action.approvedAtHeight))throw new Error('Validator governance policy was not ACTIVE at approval height');
 const scope=requiredScope(action.actionType);
 if(!policy.scope.includes(scope))throw new Error(`Validator governance policy lacks ${scope} authority`);
 const computedPolicyHash=governancePolicyHash(policy);
 if(computedPolicyHash!==args.expectedGovernancePolicyHash||action.governancePolicyHash!==computedPolicyHash)throw new Error('Validator governance policy hash does not match independently trusted policy');
 const previousHash=validatorSetStateHash(previous);const nextHash=validatorSetStateHash(next);
 if(action.previousValidatorSetHash!==previousHash||action.nextValidatorSetHash!==nextHash)throw new Error('Validator-set state hash mismatch');
 const computedPreviousRoot=validatorSetRootAtHeight(previous,action.effectiveHeight-1);
 if(computedPreviousRoot!==args.expectedPreviousValidatorSetRoot||action.previousValidatorSetRoot!==computedPreviousRoot)throw new Error('Previous validator-set root does not match independently trusted canonical root');
 const computedNextRoot=validatorSetRootAtHeight(next,action.effectiveHeight);
 if(action.nextValidatorSetRoot!==computedNextRoot)throw new Error('Next validator-set root mismatch');
 assertOnlyTargetChanged(previous,next,action.validatorId);
 const subject=action.actionType==='ROTATE_CONSENSUS_KEY'?assertConsensusKeyRotation(proof):assertMembershipTransition(proof);
 if(action.subjectIdentityHash!==validatorSubjectIdentityHash(subject))throw new Error('Validator subject identity hash mismatch');

 const eligible=policy.members.filter(member=>governanceMemberEligibleAtHeight(member,action.approvedAtHeight));
 if(eligible.length<1)throw new Error('No eligible VALIDATOR governance authority at approval height');
 const required=Math.max(policy.threshold,requiredGovernanceSupermajority(eligible.length));
 if(required>eligible.length)throw new Error(`Validator governance threshold ${required} exceeds ${eligible.length} eligible authority members`);
 const authorityById=new Map(eligible.map(member=>[member.authorityMemberId,member]));
 const payload=Buffer.from(canonicalize(action),'utf8');
 const validSigners:string[]=[];
 for(const signature of proof.signatures){
  const member=authorityById.get(signature.authorityMemberId);
  if(!member||signature.keyId!==member.keyId||signature.algorithm!=='Ed25519'||signature.domain!==VALIDATOR_SET_CHANGE_DOMAIN)continue;
  const publicKey=createPublicKey({key:Buffer.from(member.publicKeyDerB64,'base64'),format:'der',type:'spki'});
  if(verifySignature(null,payload,publicKey,Buffer.from(signature.signatureB64,'base64')))validSigners.push(member.authorityMemberId);
 }
 const uniqueValid=[...new Set(validSigners)].sort();
 if(uniqueValid.length<required)throw new Error(`Validator governance supermajority not met: ${uniqueValid.length}/${eligible.length}; ${required} required`);
 return{
  valid:true as const,
  actionId:action.actionId,
  actionType:action.actionType,
  validatorId:action.validatorId,
  approvedAtHeight:action.approvedAtHeight,
  effectiveHeight:action.effectiveHeight,
  governancePolicyHash:computedPolicyHash,
  previousValidatorSetRoot:computedPreviousRoot,
  nextValidatorSetRoot:computedNextRoot,
  previousValidatorSetHash:previousHash,
  nextValidatorSetHash:nextHash,
  eligibleAuthorityCount:eligible.length,
  requiredAuthority:required,
  validSigners:uniqueValid,
  trustedNextValidatorSet:{height:action.effectiveHeight,validatorSetRoot:computedNextRoot,stateHash:nextHash},
 };
}
